cpp/src/arrow/compute/kernels/aggregate_basic.cc

```diff
@@ -121,6 +122,168 @@ Result<std::unique_ptr<KernelState>> CountInit(KernelContext*,
+    }
+    return Status::OK();
+  }
+
+  Status MergeFrom(KernelContext*, KernelState&& src) override {
+    const auto& other_state = checked_cast<const CountDistinctImpl&>(src);
+    this->non_nulls += other_state.non_nulls;
+    this->has_nulls = this->has_nulls || other_state.has_nulls;
+    return Status::OK();
+  }
+
+  Status Finalize(KernelContext* ctx, Datum* out) override {
+    const auto& state = checked_cast<const CountDistinctImpl&>(*ctx->state());
+    const int64_t nulls = state.has_nulls ? 1 : 0;
+    switch (state.options.mode) {
+      case CountOptions::ONLY_VALID:
+        *out = Datum(state.non_nulls);
+        break;
+      case CountOptions::ALL:
+        *out = Datum(state.non_nulls + nulls);
+        break;
+      case CountOptions::ONLY_NULL:
+        *out = Datum(nulls);
+        break;
+      default:
+        DCHECK(false) << "unreachable";
+    }
+    return Status::OK();
+  }
+
+  const CountOptions options;
+  int64_t non_nulls = 0;
+  bool has_nulls = false;
+  std::unique_ptr<MemoTable> memo_table_;
+};
+
+template <typename Type, typename VisitorArgType>
+Result<std::unique_ptr<KernelState>> CountDistinctInit(KernelContext* ctx,
+                                                       const KernelInitArgs& args) {
+  return ::arrow::internal::make_unique<CountDistinctImpl<Type, VisitorArgType>>(
+      ctx->memory_pool(), static_cast<const CountOptions&>(*args.options));
+}
+
+template <typename Type, typename VisitorArgType>
+void AddCountDistinctKernel(InputType type, ScalarAggregateFunction* func) {
+  AddAggKernel(KernelSignature::Make({type}, ValueDescr::Scalar(int64())),
+               aggregate::CountDistinctInit<Type, VisitorArgType>, func);
+}
+
+template <typename Type, typename VisitorArgType = typename Type::c_type>
+struct CountDistinctKernel {
+  static void Add(InputType type, ScalarAggregateFunction* func) {
+    using PhysicalType = typename Type::PhysicalType;
+    AddCountDistinctKernel<PhysicalType, VisitorArgType>(type, func);
+  }
+};
+
+template <>
+struct CountDistinctKernel<FixedSizeBinaryType, util::string_view> {
+  static void Add(InputType type, ScalarAggregateFunction* func) {
+    AddCountDistinctKernel<FixedSizeBinaryType, util::string_view>(type, func);
+  }
+};
+
+template <>
+struct CountDistinctKernel<DecimalType, util::string_view> {
+  static void Add(InputType type, ScalarAggregateFunction* func) {
+    AddCountDistinctKernel<FixedSizeBinaryType, util::string_view>(type, func);
+  }
+};
+
+template <>
+struct CountDistinctKernel<Decimal128Type, util::string_view> {
+  static void Add(InputType type, ScalarAggregateFunction* func) {
+    AddCountDistinctKernel<FixedSizeBinaryType, util::string_view>(type, func);
+  }
+};
+
+template <>
+struct CountDistinctKernel<Decimal256Type, util::string_view> {
+  static void Add(InputType type, ScalarAggregateFunction* func) {
+    AddCountDistinctKernel<FixedSizeBinaryType, util::string_view>(type, func);
+  }
```

Comment:
The specialized `CountDistinctKernel` all have the same definition so only one would suffice. Use `enable_if` to specialize for any of these cases.

Reply:
Good catch, thanks @edponce !

Reply:
Done!